execution-core/src/transfer/moonlight.rs

```diff
@@ -338,11 +364,25 @@ impl Payload {
         let gas_price = u64::from_reader(&mut buf)?;
         let nonce = u64::from_reader(&mut buf)?;
 
-        // deserialize contract call/deploy data
-        let exec = match u8::from_reader(&mut buf)? {
+        // deserialize contract call, deploy data, or memo, if present
+        let data = match u8::from_reader(&mut buf)? {
             0 => None,
-            1 => Some(ContractExec::Call(ContractCall::from_slice(buf)?)),
-            2 => Some(ContractExec::Deploy(ContractDeploy::from_slice(buf)?)),
+            1 => Some(TransactionData::Call(ContractCall::from_slice(buf)?)),
+            2 => {
+                Some(TransactionData::Deploy(ContractDeploy::from_slice(buf)?))
+            }
+            3 => {
+                // we only build for 64-bit so this truncation is impossible
+                #[allow(clippy::cast_possible_truncation)]
+                let size = u64::from_reader(&mut buf)? as usize;
+
+                if buf.len() != size || size > MAX_MEMO_SIZE {
```

Comment:
Isn't buf.len() the whole buffer length?

Reply:
Looking at the test seems that I'm wrong...

Reply:
It is the end of the buffer, so it should indeed fill in all of the remaining bytes.